### PR TITLE
Support .NET Core 3.0

### LINKED ISSUE #118: MissingMethodException when running on mono with newer Newtonsoft.Json
This is quite an exotic setup I know, but this has my tests failing in CI, so I'll be looking into this too.

I have an example here:
https://github.com/justeat/httpclient-interception/commit/a776327d8c1de00e7ad2037e896c171c91c239f6#diff-ad2944e2db1bf6a75b8602f406da7846R13

When running under mono, if I call `.WithJsonContent(...` then it pops with a runtime exception:
```
Error Message:
   System.MissingMethodException : Method not found: JustEat.HttpClientInterception.HttpRequestInterceptionBuilder JustEat.HttpClientInterception.HttpRequestInterceptionBuilderExtensions.WithJsonContent(JustEat.HttpClientInterception.HttpRequestInterceptionBuilder,object,Newtonsoft.Json.JsonSerializerSettings)
```

In recreating this, I discovered that it's not an issue if you are referencing `Newtonsoft.Json` 9.0.1, but if you reference a higher version such as 12.x then this issue pops up.

### PR DESCRIPTION
**Pull request by martincostello**
This PR adds support for .NET Core 3.0 and some various other improvements based on the use of the .NET Core 3.0 SDK. The changes include breaking changes to the public API surface, so are versioned as 3.0.0, which is a happy coincidence for .NET Core 3.0 support.

Changes:
  * Remove Newtonsoft.Json optional parameter from `WithJsonContent()` to resolve #118.
  * Rename `Register(...)` for `byte[]` on `HttpClientInterceptorOptions` to `RegisterByteArray(...)`.
  * Rename `Register(...)` for `Stream` on `HttpClientInterceptorOptions` to `RegisterStream(...)`.
  * Remove optional `HttpMessageHandler innerHandler` parameter from `public static HttpClient CreateHttpClient(this HttpClientInterceptorOptions options, string baseAddress)`.
  * Remove support for `netstandard1.3`.
  * Add support for `net472`.
  * Use `Microsoft.NETFramework.ReferenceAssemblies` to compile `net4*` TFMs on non-Windows platforms.
  * Update SDK to .NET Core 3.0.
  * Update sample app to target ASP.NET Core 3.0.
  * Fix typos in XML documentation.
  * Add support for new System.Text.Json APIs with `WithSystemTextJsonContent()`.
  * Add new `WithNewtonsoftJsonContent()` method to replace `WithJsonContent()`.
  * Adds nullable annotations.

### CODE REVIEW

**Comment by martincostello** [comment]
Some initial benchmarks using 3.0 preview 7:

``` ini

BenchmarkDotNet=v0.11.5, OS=Windows 10.0.17134.885 (1803/April2018Update/Redstone4)
Intel Core i7-6500U CPU 2.50GHz (Skylake), 1 CPU, 4 logical and 2 physical cores
.NET Core SDK=3.0.100-preview7-012821
  [Host]     : .NET Core 3.0.0-preview7-27912-14 (CoreCLR 4.700.19.32702, CoreFX 4.700.19.36209), 64bit RyuJIT
  DefaultJob : .NET Core 3.0.0-preview7-27912-14 (CoreCLR 4.700.19.32702, CoreFX 4.700.19.36209), 64bit RyuJIT


```
|                Method |       Mean |      Error |     StdDev |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|---------------------- |-----------:|-----------:|-----------:|-------:|------:|------:|----------:|
|              GetBytes |   5.598 us |  0.2098 us |  0.5985 us | 1.6251 |     - |     - |   3.33 KB |
|               GetHtml |   5.068 us |  0.0605 us |  0.0566 us | 1.5869 |     - |     - |   3.25 KB |
| GetJsonNewtonsoftJson |  13.399 us |  0.2568 us |  0.2748 us | 4.0283 |     - |     - |   8.29 KB |
| GetJsonSystemTextJson |  10.889 us |  0.2134 us |  0.2096 us | 2.6245 |     - |     - |   5.37 KB |
|             GetStream | 565.531 us |  6.3190 us |  5.6017 us | 0.9766 |     - |     - |    3.3 KB |
|   RefitNewtonsoftJson | 396.098 us | 20.7441 us | 56.0828 us |      - |     - |     - |   17.3 KB |
|   RefitSystemTextJson |  32.784 us |  0.3224 us |  0.2858 us | 4.3945 |     - |     - |      9 KB |
```

**Comment by slang25** [comment]
RefitNewtonsoftJson vs RefitSystemTextJson! yowza

**Comment by martincostello** [comment]
I don't _quite_ believe the value for that as it differs from the baseline in the README by loads, and also these are off my work laptop which is a lesser spec compared to my own laptop which those benchmarks came from.

Will do a more like-for-like re-run at some point soon.

**Comment by slang25** [comment]
I suspected they looked a little too good, just done a run on my macbook and the numbers are looking a bit more as expected:

``` ini

BenchmarkDotNet=v0.11.5, OS=macOS Mojave 10.14.6 (18G87) [Darwin 18.7.0]
Intel Core i7-8750H CPU 2.20GHz (Coffee Lake), 1 CPU, 12 logical and 6 physical cores
.NET Core SDK=3.0.100-preview7-012821
  [Host]     : .NET Core 3.0.0-preview7-27912-14 (CoreCLR 4.700.19.32702, CoreFX 4.700.19.36209), 64bit RyuJIT
  DefaultJob : .NET Core 3.0.0-preview7-27912-14 (CoreCLR 4.700.19.32702, CoreFX 4.700.19.36209), 64bit RyuJIT


```
|                Method |      Mean |     Error |    StdDev |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|---------------------- |----------:|----------:|----------:|-------:|------:|------:|----------:|
|              GetBytes |  3.786 us | 0.0140 us | 0.0124 us | 0.7172 |     - |     - |   3.33 KB |
|               GetHtml |  5.211 us | 0.0359 us | 0.0318 us | 0.9232 |     - |     - |   4.26 KB |
| GetJsonNewtonsoftJson | 10.102 us | 0.0425 us | 0.0377 us | 1.8616 |     - |     - |   8.62 KB |
| GetJsonSystemTextJson |  7.987 us | 0.0664 us | 0.0621 us | 1.1597 |     - |     - |   5.37 KB |
|             GetStream | 36.127 us | 0.1673 us | 0.1483 us | 1.1597 |     - |     - |   5.33 KB |
|   RefitNewtonsoftJson | 30.488 us | 0.2358 us | 0.2090 us | 3.2349 |     - |     - |   15.1 KB |
|   RefitSystemTextJson | 26.002 us | 0.1760 us | 0.1561 us | 1.9531 |     - |     - |   8.99 KB |

**Comment by martincostello** [comment]
Another round from the same laptop comparing this PR running under a selection of runtimes:

## net472

``` ini

BenchmarkDotNet=v0.11.5, OS=Windows 10.0.17134.885 (1803/April2018Update/Redstone4)
Intel Core i7-6500U CPU 2.50GHz (Skylake), 1 CPU, 4 logical and 2 physical cores
  [Host]     : .NET Framework 4.7.2 (CLR 4.0.30319.42000), 64bit RyuJIT-v4.8.3801.0
  DefaultJob : .NET Framework 4.7.2 (CLR 4.0.30319.42000), 64bit RyuJIT-v4.8.3801.0


```
|                Method |      Mean |      Error |     StdDev |    Median |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|---------------------- |----------:|-----------:|-----------:|----------:|-------:|------:|------:|----------:|
|              GetBytes |  32.42 us |  1.1534 us |  3.3646 us |  31.03 us | 2.8076 |     - |     - |   5.79 KB |
|               GetHtml |  32.57 us |  0.9762 us |  1.0445 us |  32.20 us | 3.2959 |     - |     - |   6.78 KB |
| GetJsonNewtonsoftJson |  53.33 us |  0.6572 us |  0.6147 us |  53.36 us | 6.5308 |     - |     - |  13.46 KB |
| GetJsonSystemTextJson |  17.84 us |  0.3552 us |  0.6672 us |  17.62 us | 3.7231 |     - |     - |   7.63 KB |
|             GetStream | 416.05 us |  2.7719 us |  2.4572 us | 416.31 us | 2.4414 |     - |     - |   5.51 KB |
|   RefitNewtonsoftJson | 208.80 us | 24.7940 us | 72.7164 us | 206.20 us |      - |     - |     - |     24 KB |
|   RefitSystemTextJson | 184.11 us | 20.9170 us | 60.6840 us | 173.70 us |      - |     - |     - |     16 KB |


## netcoreapp2.2

``` ini

BenchmarkDotNet=v0.11.5, OS=Windows 10.0.17134.885 (1803/April2018Update/Redstone4)
Intel Core i7-6500U CPU 2.50GHz (Skylake), 1 CPU, 4 logical and 2 physical cores
.NET Core SDK=3.0.100-preview7-012821
  [Host]     : .NET Core 2.2.6 (CoreCLR 4.6.27817.03, CoreFX 4.6.27818.02), 64bit RyuJIT
  DefaultJob : .NET Core 2.2.6 (CoreCLR 4.6.27817.03, CoreFX 4.6.27818.02), 64bit RyuJIT


```
|                Method |       Mean |      Error |     StdDev |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|---------------------- |-----------:|-----------:|-----------:|-------:|------:|------:|----------:|
|              GetBytes |   5.929 us |  0.0538 us |  0.0476 us | 1.4954 |     - |     - |   3.08 KB |
|               GetHtml |   7.481 us |  0.1879 us |  0.3241 us | 1.6251 |     - |     - |   3.34 KB |
| GetJsonNewtonsoftJson |  15.492 us |  0.1194 us |  0.1117 us | 4.0894 |     - |     - |   8.42 KB |
| GetJsonSystemTextJson |  10.849 us |  0.0628 us |  0.0557 us | 2.3804 |     - |     - |   4.88 KB |
|             GetStream | 555.795 us |  5.6662 us |  4.7315 us | 1.9531 |     - |     - |   4.09 KB |
|   RefitNewtonsoftJson | 290.560 us | 11.1933 us | 31.5710 us |      - |     - |     - |  15.98 KB |
|   RefitSystemTextJson | 246.178 us | 18.9221 us | 55.4953 us |      - |     - |     - |  11.13 KB |


## netcoreapp3.0

``` ini

BenchmarkDotNet=v0.11.5, OS=Windows 10.0.17134.885 (1803/April2018Update/Redstone4)
Intel Core i7-6500U CPU 2.50GHz (Skylake), 1 CPU, 4 logical and 2 physical cores
.NET Core SDK=3.0.100-preview7-012821
  [Host]     : .NET Core 3.0.0-preview7-27912-14 (CoreCLR 4.700.19.32702, CoreFX 4.700.19.36209), 64bit RyuJIT
  DefaultJob : .NET Core 3.0.0-preview7-27912-14 (CoreCLR 4.700.19.32702, CoreFX 4.700.19.36209), 64bit RyuJIT


```
|                Method |       Mean |     Error |    StdDev |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|---------------------- |-----------:|----------:|----------:|-------:|------:|------:|----------:|
|              GetBytes |   4.030 us | 0.0310 us | 0.0274 us | 1.4648 |     - |     - |      3 KB |
|               GetHtml |   4.576 us | 0.0521 us | 0.0435 us | 1.5869 |     - |     - |   3.25 KB |
| GetJsonNewtonsoftJson |  13.884 us | 0.1830 us | 0.1622 us | 4.5166 |     - |     - |   9.27 KB |
| GetJsonSystemTextJson |   8.231 us | 0.0791 us | 0.0740 us | 2.3041 |     - |     - |   4.71 KB |
|             GetStream | 590.482 us | 8.2968 us | 7.7608 us | 0.9766 |     - |     - |   3.65 KB |
|   RefitNewtonsoftJson |  36.762 us | 0.7259 us | 1.2903 us | 7.2021 |     - |     - |  14.77 KB |
|   RefitSystemTextJson |  35.792 us | 0.9099 us | 2.5959 us | 4.6387 |     - |     - |   9.65 KB |

File: tests/HttpClientInterception.Benchmarks/InterceptionBenchmarks.cs [hunk 1]
```diff
@@ -155,7 +155,11 @@ public async Task<HttpContent> SerializeAsync<T>(T item)
                 }
                 catch (Exception)
                 {
+#if NETCOREAPP3_0
                     await stream.DisposeAsync();
+#else
+                    stream.Dispose();
+#endif
```
Comment (slang25) [inline tests/HttpClientInterception.Benchmarks/InterceptionBenchmarks.cs]: fwiw `MemoryStream` `DisposeAsync` just calls `Dispose`

Comment (martincostello) [reply tests/HttpClientInterception.Benchmarks/InterceptionBenchmarks.cs]: Maybe I'll simplify it to just use the sync one everywhere as it's always a concrete `MemoryStream`.

**Comment by martincostello** [comment]
Here's the benchmarks from my own laptop:

## net472

``` ini

BenchmarkDotNet=v0.11.5, OS=Windows 10.0.18362
Intel Core i7-6700HQ CPU 2.60GHz (Skylake), 1 CPU, 8 logical and 4 physical cores
  [Host]     : .NET Framework 4.7.2 (CLR 4.0.30319.42000), 64bit RyuJIT-v4.8.3815.0
  DefaultJob : .NET Framework 4.7.2 (CLR 4.0.30319.42000), 64bit RyuJIT-v4.8.3815.0


```
|                Method |      Mean |      Error |     StdDev |    Median |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|---------------------- |----------:|-----------:|-----------:|----------:|-------:|------:|------:|----------:|
|              GetBytes |  23.34 us |  0.7669 us |  2.2249 us |  22.98 us | 1.8616 |     - |     - |   5.78 KB |
|               GetHtml |  25.47 us |  0.5075 us |  1.0253 us |  25.39 us | 2.1973 |     - |     - |   6.77 KB |
| GetJsonNewtonsoftJson |  55.30 us |  5.9576 us | 17.5660 us |  47.37 us | 4.3335 |     - |     - |  13.45 KB |
| GetJsonSystemTextJson |  18.60 us |  0.6740 us |  1.9873 us |  18.07 us | 2.4719 |     - |     - |   7.63 KB |
|             GetStream |  63.51 us |  0.9484 us |  0.7920 us |  63.42 us | 1.7090 |     - |     - |   5.51 KB |
|   RefitNewtonsoftJson |  48.96 us |  1.3848 us |  3.8603 us |  48.68 us | 5.7983 |     - |     - |   17.9 KB |
|   RefitSystemTextJson | 184.56 us | 21.2091 us | 61.1930 us | 175.25 us |      - |     - |     - |     16 KB |


## netcoreapp2.2

``` ini

BenchmarkDotNet=v0.11.5, OS=Windows 10.0.18362
Intel Core i7-6700HQ CPU 2.60GHz (Skylake), 1 CPU, 8 logical and 4 physical cores
.NET Core SDK=3.0.100-preview7-012821
  [Host]     : .NET Core 2.2.6 (CoreCLR 4.6.27817.03, CoreFX 4.6.27818.02), 64bit RyuJIT
  DefaultJob : .NET Core 2.2.6 (CoreCLR 4.6.27817.03, CoreFX 4.6.27818.02), 64bit RyuJIT


```
|                Method |      Mean |     Error |    StdDev |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|---------------------- |----------:|----------:|----------:|-------:|------:|------:|----------:|
|              GetBytes |  6.796 us | 0.1164 us | 0.1032 us | 1.2131 |     - |     - |   3.75 KB |
|               GetHtml |  8.350 us | 0.1561 us | 0.1533 us | 1.4191 |     - |     - |   4.37 KB |
| GetJsonNewtonsoftJson | 15.753 us | 0.2489 us | 0.2328 us | 2.9297 |     - |     - |   9.09 KB |
| GetJsonSystemTextJson | 11.882 us | 0.2317 us | 0.2054 us | 1.8005 |     - |     - |   5.55 KB |
|             GetStream | 58.654 us | 1.1585 us | 1.2396 us | 1.4038 |     - |     - |   4.45 KB |
|   RefitNewtonsoftJson | 44.093 us | 0.8799 us | 1.7775 us | 4.9438 |     - |     - |  15.33 KB |
|   RefitSystemTextJson | 39.321 us | 0.7748 us | 0.8291 us | 3.2959 |     - |     - |  10.14 KB |


## netcoreapp3.0

``` ini

BenchmarkDotNet=v0.11.5, OS=Windows 10.0.18362
Intel Core i7-6700HQ CPU 2.60GHz (Skylake), 1 CPU, 8 logical and 4 physical cores
.NET Core SDK=3.0.100-preview7-012821
  [Host]     : .NET Core 3.0.0-preview7-27912-14 (CoreCLR 4.700.19.32702, CoreFX 4.700.19.36209), 64bit RyuJIT
  DefaultJob : .NET Core 3.0.0-preview7-27912-14 (CoreCLR 4.700.19.32702, CoreFX 4.700.19.36209), 64bit RyuJIT


```
|                Method |      Mean |     Error |    StdDev |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|---------------------- |----------:|----------:|----------:|-------:|------:|------:|----------:|
|              GetBytes |  3.862 us | 0.0647 us | 0.0605 us | 0.9766 |     - |     - |      3 KB |
|               GetHtml |  5.237 us | 0.1046 us | 0.1598 us | 1.2665 |     - |     - |   3.92 KB |
| GetJsonNewtonsoftJson | 12.979 us | 0.3346 us | 0.3286 us | 3.0212 |     - |     - |   9.27 KB |
| GetJsonSystemTextJson |  9.864 us | 0.1930 us | 0.2706 us | 1.8311 |     - |     - |    5.7 KB |
|             GetStream | 55.825 us | 1.0538 us | 1.1713 us | 0.9766 |     - |     - |    3.3 KB |
|   RefitNewtonsoftJson | 33.374 us | 0.6640 us | 0.9089 us | 4.7607 |     - |     - |  14.77 KB |
|   RefitSystemTextJson | 29.938 us | 0.2313 us | 0.1932 us | 2.9297 |     - |     - |   9.32 KB |

**Comment by martincostello** [comment]
And master with the latest code:

``` ini

BenchmarkDotNet=v0.11.5, OS=Windows 10.0.18362
Intel Core i7-6700HQ CPU 2.60GHz (Skylake), 1 CPU, 8 logical and 4 physical cores
.NET Core SDK=2.2.401
  [Host]     : .NET Core 2.2.6 (CoreCLR 4.6.27817.03, CoreFX 4.6.27818.02), 64bit RyuJIT
  DefaultJob : .NET Core 2.2.6 (CoreCLR 4.6.27817.03, CoreFX 4.6.27818.02), 64bit RyuJIT


```
|    Method |      Mean |     Error |   StdDev |    Median |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|---------- |----------:|----------:|---------:|----------:|-------:|------:|------:|----------:|
|  GetBytes |  8.134 us | 0.4851 us | 1.423 us |  7.610 us | 1.0986 |     - |     - |   3.41 KB |
|   GetHtml |  7.998 us | 0.3588 us | 1.006 us |  7.761 us | 1.0834 |     - |     - |   3.34 KB |
|   GetJson | 19.742 us | 0.8795 us | 2.466 us | 18.840 us | 3.1433 |     - |     - |   9.73 KB |
| GetStream | 67.553 us | 1.7217 us | 4.940 us | 66.936 us | 1.3428 |     - |     - |   4.45 KB |
|     Refit | 42.700 us | 0.8524 us | 2.333 us | 41.962 us | 4.9438 |     - |     - |  15.29 KB |

**Comment by martincostello** [comment]
Looks like the AzDO CI for Linux has a weird race condition on the file system that's making the build fail, I'll fix it later somehow. Probably an unintended side-effect of adding the 4th TFM.

**Pull request merged**